### ISSUE TITLE
HTTP access logging can break asynchronous servlets (and maybe asynchronous JAX-RS services too)

### ISSUE THREAD

**Issue by jo-ka** [payara/Payara#1401]
# Description #

----------

There's a recent open bug report for GlassFish [here](https://java.net/jira/browse/GLASSFISH-21651) that points out that the log entries created for asynchronous servlets are wrong. However, I have strong reason to believe that under certain circumstances, this broken logging can even lead to the failure of the execution of asynchronous servlets itself, which manifests itself in seemingly random, race-condition-like exceptions.

## Expected Outcome

With regards to the log entries: correct HTTP status code and message size. With regards to the execution of asynchronous servlets: deterministically successful execution.

## Current Outcome

(Note: I'm quoting exception stack traces from GF 4.1 here, but I could reproduce the bug on Payara 4.1.1.164 in principle, so apart from potentially wrong line numbers, they should still apply)

As noted in the linked java.net thread, access log records for asynchronous servlets (usually) have their status code set to 200 and their message size to 0, e.g.

`"10.0.30.83" "08/Feb/2017:08:58:42 +0100" "GET [...] HTTP/1.1" 200 0`

The much bigger problem are exceptions like shown below though, which were observed with a frequency of roughly 1 in 10000 requests.

    [2016-12-15T14:32:56.550+0100] [glassfish 4.1] [SEVERE] [] [javax.enterprise.web.core] [jk-connector(11)] [timeMillis: 1481808776550] [] [[
    An exception or error occurred in the container during the request processing
    java.lang.IllegalStateException: Current state = RESET, new state = FLUSHED
        at java.nio.charset.CharsetDecoder.throwIllegalStateException(CharsetDecoder.java:968)
        at java.nio.charset.CharsetDecoder.flush(CharsetDecoder.java:657)
        at java.nio.charset.CharsetDecoder.decode(CharsetDecoder.java:786)
        at java.nio.charset.Charset.decode(Charset.java:806)
        at org.glassfish.grizzly.memory.HeapBuffer.toStringContent(HeapBuffer.java:859)
        at org.glassfish.grizzly.http.util.BufferChunk.toString(BufferChunk.java:497)
        at org.glassfish.grizzly.http.util.DataChunk.toString(DataChunk.java:291)
        at org.glassfish.grizzly.http.util.DataChunk.toString(DataChunk.java:283)
        at org.glassfish.grizzly.http.HttpRequestPacket.getQueryString(HttpRequestPacket.java:251)
        at org.glassfish.grizzly.http.server.Request.getQueryString(Request.java:1697)
        at org.apache.catalina.connector.Request.getQueryString(Request.java:2742)
        at org.apache.catalina.connector.RequestFacade.getQueryString(RequestFacade.java:824)
        at com.sun.enterprise.web.accesslog.DefaultAccessLogFormatterImpl.appendRequestInfo(DefaultAccessLogFormatterImpl.java:536)
        at com.sun.enterprise.web.accesslog.DefaultAccessLogFormatterImpl.appendLogEntry(DefaultAccessLogFormatterImpl.java:288)
        at com.sun.enterprise.web.PEAccessLogValve.postInvoke(PEAccessLogValve.java:648)
        [...]
        
    [2016-12-15T12:56:34.020+0100] [glassfish 4.1] [SEVERE] [] [javax.enterprise.web.core] [http-thread-pool(7)] [timeMillis: 1481802994020] [] [[
    An exception or error occurred in the container during the request processing
    java.lang.IllegalArgumentException
        at java.nio.Buffer.position(Buffer.java:236)
        at org.glassfish.grizzly.memory.Buffers.setPositionLimit(Buffers.java:292)
        at org.glassfish.grizzly.memory.HeapBuffer.toStringContent(HeapBuffer.java:867)
        at org.glassfish.grizzly.http.util.BufferChunk.toString(BufferChunk.java:502)
        at org.glassfish.grizzly.http.util.DataChunk.toString(DataChunk.java:298)
        at org.glassfish.grizzly.http.server.Request.getQueryString(Request.java:1735)
        at org.apache.catalina.connector.Request.getQueryString(Request.java:2742)
        at org.apache.catalina.connector.RequestFacade.getQueryString(RequestFacade.java:824)
        at com.sun.enterprise.web.accesslog.DefaultAccessLogFormatterImpl.appendRequestInfo(DefaultAccessLogFormatterImpl.java:536)
        at com.sun.enterprise.web.accesslog.DefaultAccessLogFormatterImpl.appendLogEntry(DefaultAccessLogFormatterImpl.java:288)
        at com.sun.enterprise.web.PEAccessLogValve.postInvoke(PEAccessLogValve.java:648)
        at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:749)
        at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:673)
        [...]
        
    [2016-12-16T13:13:21.220+0100] [glassfish 4.1] [SEVERE] [] [javax.enterprise.web.core] [jk-connector(11)] [timeMillis: 1481890401220] [] [[
    An exception or error occurred in the container during the request processing
    java.lang.NullPointerException
        at java.nio.CharBuffer.put(CharBuffer.java:915)
        at com.sun.enterprise.web.accesslog.DefaultAccessLogFormatterImpl.appendRequestInfo(DefaultAccessLogFormatterImpl.java:538)
        [...]

This last one was particularly revealing, as the source code of DefaultAccessLogFormatterImpl around line 538 reads:

    if (hreq.getQueryString() != null) {
        cb.put('?');
        cb.put(hreq.getQueryString());
    }

So obviously another thread must have led to getQueryString() returning a non-null result at first and then null only moments later.

## Steps to reproduce

Enter asynchronous servlet execution by issuing AsyncContext.start(Runnable r), access the query string in the Runnable code using HttpServletRequest.getQueryString(), make sure HTTP access logging is enabled, and send 10,000s of requests.

### Samples

outline:

    protected void service( HttpServletRequest servletRequest, HttpServletResponse servletResponse ) {
        final AsyncContext aContext = servletRequest.startAsync( servletRequest, servletResponse );
        Runnable asyncWrapper = new Runnable() {
            public void run() {
                doSomethingWith( servletRequest.getQueryString() );
            }
        }
        aContext.start( asyncWrapper );
    }

## Context

If I go searching for this kind of errors on the web, I find the recurring recommendation not to keep the request (or response) in instance variables of the servlet, not to modify the request or response on different threads etc. All of which I made sure to follow.

As I had noticed that all the exceptions had something about the query string in their stack trace, and in fact there was also one occasion where a debug log output in my asynchronous thread showed a query string that seemed to be garbled from two separate requests, I directed my attention to this.

So I added a log output of the query string on the request-processing thread, too. Maybe surprisingly, the mere presence of that logging statement led to the eradication of all those random exceptions.

Through debugging server code and after having read about what Tomcat and Jetty do, I realized request objects are recycled in Grizzly. I also found that a lot of "magic" happens when one accesses `HttpServletRequest.getQueryString()` for the first time, i.e. DataChunks are created, buffers emptied/filled, values cached etc.

So in summary this looks to me as if the concurrent access of the buffers underlying the query string is what causes this bug. I say it's a bug because I believe I have not violated any contracts by merely accessing a supposedly read-only value on a different thread. There's that bug for sure that the access log records are apparently written too early, i.e. before AsyncContext.complete() was called, reported on the linked thread. Accessing the query string leads to race conditions, but maybe there are even more consequences if other state of the request is accessed on another thread while the access logging already tries to write the record.

I've put "and maybe asynchronous JAX-RS services too" in the title of the issue because I have also rarely witnessed a random `java.lang.IllegalStateException: Not Suspended` in a completely different module where I use a JAX-RS method with a `@Suspended AsyncResponse asyncResponse` parameter. I did not get around to testing this extensively yet. A simple method to test if this is also related to the logging would be to completely disable the HTTP access log, which I sucessfully tried for the asynchronous servlets (i.e. no more exceptions then).

## Environment ##

- **Payara Version**: 4.1.1.164
- **Edition**: Full
- **JDK Version**: OpenJDK 8u111
- **Operating System**: Linux

**Comment by fturizo** [comment]
Thank you for this report @jo-ka. 

I could reproduce the issue with the access logging incorrect values for the HTTP status code and payload size. I created the internal JIRA Issue `PAYARA-1585` to work on this. However, I couldn't reproduce the exceptions you are getting with a sample asynchronous servlet using high volume of requests (I tested the server under heavy load using **JMeter** and increased the number to 1M of requests and got no errors reported), so maybe that issue is related to some specific logic used inside the task executed in the separated thread.
